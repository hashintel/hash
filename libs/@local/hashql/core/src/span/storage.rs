use orx_concurrent_vec::ConcurrentVec;

use super::{Span, SpanId, entry::Entry};

/// A collection of spans within a single source.
///
/// The storage is append-only and does not support the removal or modification of spans.
/// Once inserted spans are to be considered immutable and can be referred to via their `SpanId`,
/// which is returned on insertion.
#[derive(Debug)]
pub struct SpanStorage<S> {
    inner: ConcurrentVec<S>,
}

impl<S> SpanStorage<S> {
    #[must_use]
    pub fn new() -> Self {
        Self {
            inner: ConcurrentVec::new(),
        }
    }
}

impl<S> SpanStorage<S>
where
    S: Span,
{
    #[expect(
        clippy::cast_possible_truncation,
        reason = "The arena is not expected to be larger than u32::MAX + debug assertions"
    )]
    pub fn insert(&self, span: S) -> SpanId {
        const MAX_LEN: usize = u32::MAX as usize;

        let index = self.inner.push(span);

        // The `as` here is safe, because if we're at `u32::MAX` elements, the next push would
        // overflow.
        debug_assert!(index <= MAX_LEN, "Arena is full");

        SpanId::new(index as u32)
    }

    pub fn modify(&self, span: SpanId, func: impl FnMut(&mut S)) -> bool {
        let index = span.value() as usize;

        let Some(element) = self.inner.get(index) else {
            return false;
        };

        element.update(func);
        true
    }

    #[must_use]
    pub fn get(&self, span: SpanId) -> Option<Entry<'_, S>> {
        let index = span.value() as usize;

        self.inner.get(index).map(Entry::new)
    }

    #[must_use]
    pub fn get_cloned(&self, span: SpanId) -> Option<S>
    where
        S: Clone,
    {
        let index = span.value() as usize;

        self.inner.get_cloned(index)
    }

    /// Returns a vector of ancestor span IDs for the given span.
    ///
    /// # Panics
    ///
    /// This function will panic if a circular reference is detected in the span hierarchy.
    pub fn ancestors(&self, span: SpanId) -> Vec<SpanId> {
        let mut visited = vec![span];
        let mut ancestors = Vec::new();

        let mut current = self.get(span);

        while let Some(entry) = current.take() {
            let parent_id = entry.map(Span::parent_id);

            if let Some(parent_id) = parent_id {
                assert!(
                    !visited.contains(&parent_id),
                    "circular span reference detected"
                );
                visited.push(parent_id);

                ancestors.push(parent_id);
                current = self.get(parent_id);
            }
        }

        ancestors
    }
}

impl<S> Default for SpanStorage<S> {
    fn default() -> Self {
        Self::new()
    }
}
