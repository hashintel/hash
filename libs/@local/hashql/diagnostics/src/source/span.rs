use core::fmt::Display;

use text_size::TextRange;

use super::SourceId;

/// A trait for types that represent spans within source code for diagnostic purposes.
///
/// [`DiagnosticSpan`] provides a common interface for different span representations
/// used throughout the diagnostic system. Spans identify specific locations or ranges
/// within source files and can be resolved to absolute positions for rendering.
///
/// The trait supports hierarchical spans where a span may have ancestors that provide
/// additional context or offset information.
///
/// # Examples
///
/// ```
/// use core::fmt;
///
/// use hashql_diagnostics::source::{DiagnosticSpan, SourceId};
/// use text_size::TextRange;
///
/// #[derive(Debug)]
/// struct SimpleSpan {
///     source_id: SourceId,
///     range: TextRange,
/// }
///
/// impl fmt::Display for SimpleSpan {
///     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
///         write!(f, "{}:{:?}", self.source_id, self.range)
///     }
/// }
///
/// impl<R> DiagnosticSpan<R> for SimpleSpan {
///     fn source(&self) -> SourceId {
///         self.source_id
///     }
///
///     fn span(&self, _resolver: &mut R) -> Option<TextRange> {
///         Some(self.range)
///     }
///
///     fn ancestors(&self, _resolver: &mut R) -> impl IntoIterator<Item = Self> + use<R> {
///         [] // No ancestors for this simple span
///     }
/// }
/// ```
pub trait DiagnosticSpan<R>: Display {
    /// Returns the source file identifier for this span.
    ///
    /// The returned [`SourceId`] identifies which source file this span
    /// references within a [`Sources`] collection.
    ///
    /// [`Sources`]: super::Sources
    fn source(&self) -> SourceId;

    /// Resolves this span to a text range within its source file.
    ///
    /// Returns `None` if the span cannot be resolved, which may occur if
    /// the span references invalid locations or if the resolver cannot
    /// process this particular span type.
    fn span(&self, resolver: &mut R) -> Option<TextRange>;

    /// Returns ancestor spans that provide additional context for this span.
    ///
    /// Ancestors are used to build hierarchical span relationships, where
    /// the final resolved position may depend on the positions of parent
    /// spans. For example, a span within a macro expansion might have
    /// the macro invocation site as an ancestor.
    ///
    /// The returned spans are processed in order, with each ancestor's
    /// position contributing to the final resolved location.
    #[expect(clippy::min_ident_chars, reason = "false-positive")]
    fn ancestors(&self, resolver: &mut R) -> impl IntoIterator<Item = Self> + use<Self, R>;
}

/// An absolute span representing a resolved location within a source file.
///
/// [`SourceSpan`] represents a span that has been fully resolved to an absolute
/// position within a specific source file. Unlike diagnostic spans that may need
/// resolution, source spans contain concrete location information that can be
/// directly used for rendering and display.
///
/// # Examples
///
/// ```
/// use hashql_diagnostics::source::{SourceId, SourceSpan};
/// use text_size::TextRange;
///
/// let source_id = SourceId::new_unchecked(0);
/// let range = TextRange::new(10.into(), 15.into());
/// let span = SourceSpan::from_parts(source_id, range);
///
/// assert_eq!(span.source(), source_id);
/// assert_eq!(span.range(), range);
/// ```
#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub struct SourceSpan {
    source: SourceId,
    range: TextRange,
}

impl SourceSpan {
    /// Resolves a diagnostic span to an absolute source span.
    ///
    /// This method takes a diagnostic span and resolves it to an absolute position
    /// within its source file by combining the span's own range with the ranges
    /// of all its ancestors. This process handles hierarchical spans where the
    /// final position depends on multiple nested contexts.
    ///
    /// Returns `None` if the span or any of its ancestors cannot be resolved.
    ///
    /// # Examples
    ///
    /// ```
    /// # use core::fmt;
    /// # use text_size::TextRange;
    /// # use hashql_diagnostics::source::{DiagnosticSpan, SourceId, SourceSpan};
    /// # #[derive(Debug)]
    /// # struct SimpleSpan { source_id: SourceId, range: TextRange }
    /// # impl fmt::Display for SimpleSpan {
    /// #     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    /// #         write!(f, "{}:{:?}", self.source_id, self.range)
    /// #     }
    /// # }
    /// # impl<R> DiagnosticSpan<R> for SimpleSpan {
    /// #     fn source(&self) -> SourceId { self.source_id }
    /// #     fn span(&self, _: &mut R) -> Option<TextRange> { Some(self.range) }
    /// #     fn ancestors(&self, _: &mut R) -> impl IntoIterator<Item = Self> + use<R> { [] }
    /// # }
    /// # struct DummyResolver;
    ///
    /// let source_id = SourceId::new_unchecked(0);
    /// let diagnostic_span = SimpleSpan {
    ///     source_id,
    ///     range: TextRange::new(10.into(), 20.into()),
    /// };
    ///
    /// let mut resolver = DummyResolver;
    /// if let Some(absolute_span) = SourceSpan::resolve(&diagnostic_span, &mut resolver) {
    ///     assert_eq!(absolute_span.source(), source_id);
    /// }
    /// ```
    pub fn resolve<S, R>(span: &S, resolver: &mut R) -> Option<Self>
    where
        S: DiagnosticSpan<R>,
    {
        let source = span.source();

        let mut range = span.span(resolver)?;

        for ancestor in span.ancestors(resolver) {
            range += ancestor.span(resolver)?.start();
        }

        Some(Self { source, range })
    }

    /// Creates a new source span from a source ID and text range.
    ///
    /// # Examples
    ///
    /// ```
    /// use hashql_diagnostics::source::{SourceId, SourceSpan};
    /// use text_size::TextRange;
    ///
    /// let source_id = SourceId::new_unchecked(1);
    /// let range = TextRange::new(5.into(), 10.into());
    /// let span = SourceSpan::from_parts(source_id, range);
    /// ```
    #[must_use]
    pub const fn from_parts(source: SourceId, range: TextRange) -> Self {
        Self { source, range }
    }

    /// Returns the source file identifier for this span.
    #[must_use]
    pub const fn source(&self) -> SourceId {
        self.source
    }

    /// Returns the text range within the source file.
    #[must_use]
    pub const fn range(&self) -> TextRange {
        self.range
    }
}
