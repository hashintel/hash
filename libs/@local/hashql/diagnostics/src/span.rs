use core::fmt::Display;

use error_stack::Report;
use text_size::{TextRange, TextSize};

use crate::error::ResolveError;

pub trait DiagnosticSpan<Context>: Display {
    fn span(&self, context: &mut Context) -> Option<TextRange>;

    fn ancestors(
        &self,
        context: &mut Context,
    ) -> impl IntoIterator<Item = Self> + use<Self, Context>;
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub struct AbsoluteDiagnosticSpan {
    range: TextRange,
}

impl AbsoluteDiagnosticSpan {
    /// Creates a new `AbsoluteDiagnosticSpan` from a diagnostic span and its context.
    ///
    /// # Errors
    ///
    /// Returns `ResolveError::UnknownSpan` if either the span or any of its ancestors
    /// cannot be resolved in the provided context.
    pub fn new<S, C>(span: &S, context: &mut C) -> Result<Self, Report<ResolveError>>
    where
        S: DiagnosticSpan<C>,
    {
        let mut range = span
            .span(context)
            .ok_or_else(|| ResolveError::UnknownSpan {
                span: span.to_string(),
            })?;

        for ancestor in span.ancestors(context) {
            range += ancestor
                .span(context)
                .ok_or_else(|| ResolveError::UnknownSpan {
                    span: span.to_string(),
                })?
                .start();
        }

        Ok(Self { range })
    }

    #[must_use]
    pub const fn from_range(range: TextRange) -> Self {
        Self { range }
    }

    #[must_use]
    pub const fn range(self) -> TextRange {
        self.range
    }

    pub(crate) const fn full() -> Self {
        Self {
            range: TextRange::new(TextSize::new(0), TextSize::new(u32::MAX)),
        }
    }
}

impl ariadne::Span for AbsoluteDiagnosticSpan {
    type SourceId = ();

    fn source(&self) -> &Self::SourceId {
        &()
    }

    fn start(&self) -> usize {
        self.range.start().into()
    }

    fn end(&self) -> usize {
        self.range.end().into()
    }

    fn len(&self) -> usize {
        self.range.len().into()
    }

    fn is_empty(&self) -> bool {
        self.range.is_empty()
    }

    #[expect(
        clippy::cast_possible_truncation,
        reason = "Text will never be larger than u32::MAX (4GiB) due to the use of `TextSize`"
    )]
    fn contains(&self, offset: usize) -> bool {
        self.range.contains(TextSize::from(offset as u32))
    }
}
