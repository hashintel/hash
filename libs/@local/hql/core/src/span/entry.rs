use orx_concurrent_vec::ConcurrentElement;

/// A read-only entry into a concurrent element.
pub struct Entry<'a, S>(&'a ConcurrentElement<S>);

impl<'a, S> Entry<'a, S> {
    pub(crate) const fn new(element: &'a ConcurrentElement<S>) -> Self {
        Entry(element)
    }

    #[must_use]
    pub fn cloned(&self) -> S
    where
        S: Clone,
    {
        self.0.cloned()
    }

    #[must_use]
    pub fn copied(&self) -> S
    where
        S: Copy,
    {
        self.0.copied()
    }

    pub fn map<T>(&self, func: impl FnOnce(&S) -> T) -> T {
        self.0.map(func)
    }
}
