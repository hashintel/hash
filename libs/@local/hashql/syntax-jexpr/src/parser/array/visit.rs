use hashql_core::span::SpanId;
use hashql_diagnostics::Diagnostic;
use text_size::TextRange;

use crate::{
    ParserState,
    error::ResultExt as _,
    lexer::{syntax_kind::SyntaxKind, syntax_kind_set::SyntaxKindSet, token::Token},
    parser::{
        array::error::{
            ArrayDiagnosticCategory, consecutive_commas, leading_commas, trailing_commas,
        },
        complex::{VerifyState, verify_no_repeat},
        error::unexpected_token,
    },
    span::Span,
};

const EXPECTED_ARRAY_SEP: SyntaxKindSet =
    SyntaxKindSet::from_slice(&[SyntaxKind::Comma, SyntaxKind::RBracket]);

/// Parse an array from the lexer
///
/// Assumes that the lexer has already consumed the opening bracket.
///
/// # Panics
///
/// Panics if the lexer has not consumed the opening bracket.
#[expect(
    clippy::needless_pass_by_value,
    reason = "API contract, we want to signify to the user, we're now proceeding with this \
              specific token. Not that we hold it temporary, but instead that we consume it."
)]
pub(crate) fn visit_array<'arena, 'source, C>(
    state: &mut ParserState<'arena, 'source>,
    token: Token<'source>,
    mut on_item: impl FnMut(&mut ParserState<'arena, 'source>) -> Result<(), Diagnostic<C, SpanId>>,
) -> Result<TextRange, Diagnostic<C, SpanId>>
where
    C: From<ArrayDiagnosticCategory>,
{
    debug_assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

    let mut span = token.span;
    let mut index: usize = 0;

    loop {
        let next = state
            .peek_required()
            .change_category(ArrayDiagnosticCategory::Lexer)
            .change_category(C::from)?;

        let (next_kind, next_span) = (next.kind.syntax(), next.span);

        if next_kind == SyntaxKind::RBracket {
            state
                .advance()
                .change_category(ArrayDiagnosticCategory::Lexer)
                .change_category(C::from)?;

            span = span.cover(next_span);
            break;
        }

        if index == 0 {
            verify_no_repeat(
                state,
                SyntaxKindSet::from_slice(&[SyntaxKind::Comma]),
                SyntaxKindSet::EMPTY,
                |_, spans, _| leading_commas(&spans),
            )
            .change_category(C::from)?;
        } else if index != 0 {
            // we need to check if the next token is a comma
            // in case it isn't we error out
            if next_kind == SyntaxKind::Comma {
                // advance the cursor to the next token and continue as it nothing has happened
                state
                    .advance()
                    .change_category(ArrayDiagnosticCategory::Lexer)
                    .change_category(C::from)?;

                verify_no_repeat(
                    state,
                    SyntaxKindSet::from_slice(&[SyntaxKind::Comma]),
                    SyntaxKindSet::from_slice(&[SyntaxKind::RBracket]),
                    |state, mut spans, verify| match verify {
                        VerifyState::Trailing => {
                            // if trailing comma is found, then the first comma is also affected
                            spans.insert(0, state.insert_range(next_span));
                            trailing_commas(&spans)
                        }
                        VerifyState::Consecutive => consecutive_commas(&spans),
                    },
                )
                .change_category(C::from)?;
            } else {
                let span = state.insert_span(Span {
                    range: next_span,
                    pointer: Some(state.current_pointer()),
                    parent_id: None,
                });

                // do not consume the token, so that we can do recoverable parsing (in the future)
                return Err(unexpected_token(
                    span,
                    ArrayDiagnosticCategory::ExpectedSeparator,
                    EXPECTED_ARRAY_SEP,
                )
                .map_category(C::from));
            }
        }

        state.enter(jsonptr::Token::from(index), |state| on_item(state))?;
        index += 1;
    }

    Ok(span)
}

#[cfg(test)]
mod tests {
    use insta::{assert_snapshot, with_settings};
    use text_size::TextSize;

    use crate::{
        ParserState,
        error::ResultExt as _,
        lexer::{error::unexpected_eof, syntax_kind::SyntaxKind},
        parser::{
            array::{
                error::{ArrayDiagnostic, ArrayDiagnosticCategory},
                visit::visit_array,
            },
            test::{bind_context, bind_state},
        },
        span::Span,
        test::render_diagnostic,
    };

    #[test]
    fn empty_array() {
        bind_context!(let context = "[]");
        bind_state!(let mut state from context);

        // Get the first token which should be the opening bracket
        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let mut items_count = 0;
        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |_| {
            // This callback should not be called for an empty array
            items_count += 1;
            Ok(())
        });

        let range = result.expect("should be able to parse empty array");
        assert_eq!(range.start(), TextSize::new(0));
        assert_eq!(range.end(), TextSize::new(2));

        assert_eq!(items_count, 0);
    }

    #[test]
    fn single_item_array() {
        bind_context!(let context = "[42]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let mut items_count = 0;
        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |state| {
            let token = state.advance().expect("should have at least one token");
            assert_eq!(token.kind.syntax(), SyntaxKind::Number);
            items_count += 1;
            Ok(())
        });

        let range = result.expect("should be able to parse single item array");
        assert_eq!(range.start(), TextSize::new(0));
        assert_eq!(range.end(), TextSize::new(4));
        assert_eq!(items_count, 1);
    }

    #[test]
    fn multiple_item_array() {
        bind_context!(let context = "[1, 2, 3]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let mut items = Vec::new();
        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |state| {
            let token = state.advance().expect("should have at least one token");
            items.push(token.kind.syntax());
            Ok(())
        });

        let range = result.expect("should be able to parse multiple item array");
        assert_eq!(range.start(), TextSize::new(0));
        assert_eq!(range.end(), TextSize::new(9));
        assert_eq!(
            items,
            [SyntaxKind::Number, SyntaxKind::Number, SyntaxKind::Number]
        );
    }

    #[test]
    fn array_with_different_types() {
        bind_context!(let context = "[1, \"hello\", true, null]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let mut items = Vec::new();
        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |state| {
            let token = state.advance().expect("should have at least one token");
            items.push(token.kind.syntax());
            Ok(())
        });

        let range = result.expect("should be able to parse array with different types");
        assert_eq!(range.start(), TextSize::new(0));
        assert_eq!(range.end(), TextSize::new(24));
        assert_eq!(
            items,
            [
                SyntaxKind::Number,
                SyntaxKind::String,
                SyntaxKind::True,
                SyntaxKind::Null
            ]
        );
    }

    #[test]
    fn trailing_comma() {
        bind_context!(let context = "[1, 2,]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let mut callback_count = 0;
        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |state| {
            state.advance().expect("should have at least one token");

            callback_count += 1;
            Ok(())
        });

        let diagnostic = result.expect_err("should not be able to parse array with trailing comma");
        let report = render_diagnostic(context.input, diagnostic, &context.spans);

        with_settings!({
            description => "Array with trailing comma should fail"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, report, &context.input);
        });

        assert_eq!(callback_count, 2);
    }

    #[test]
    fn missing_comma() {
        bind_context!(let context = "[1 2]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let mut callback_count = 0;
        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |state| {
            state.advance().expect("should have at least one token");
            callback_count += 1;
            Ok(())
        });

        let diagnostic = result.expect_err("should not be able to parse array with missing comma");
        let report = render_diagnostic(context.input, diagnostic, &context.spans);

        with_settings!({
            description => "Array with missing comma should fail"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, report, &context.input);
        });

        assert_eq!(callback_count, 1);
    }

    #[test]
    fn error_in_callback() {
        bind_context!(let context = "[1, 2, 3]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let mut callback_count = 0;
        let result = visit_array(&mut state, token, |state| {
            let token = state.advance().expect("should have at least one token");
            callback_count += 1;

            if callback_count == 2 {
                return Err(unexpected_eof(state.insert_span(Span {
                    range: token.span,
                    pointer: None,
                    parent_id: None,
                })))
                .change_category(ArrayDiagnosticCategory::Lexer);
            }

            Ok(())
        });

        assert!(result.is_err(), "should propagate error from callback");
        assert_eq!(
            callback_count, 2,
            "callback should be called twice before error"
        );
    }

    #[test]
    fn unclosed_array() {
        bind_context!(let context = "[1, 2");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |state| {
            state.advance().expect("should have at least one token");
            Ok(())
        });

        let diagnostic = result.expect_err("should fail with unclosed array");
        let report = render_diagnostic(context.input, diagnostic, &context.spans);

        with_settings!({
            description => "Unclosed array should fail"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, report, &context.input);
        });
    }

    #[test]
    fn empty_array_with_whitespace() {
        bind_context!(let context = "[   ]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let mut items_count = 0;
        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |_| {
            items_count += 1;
            Ok(())
        });

        let range = result.expect("should be able to parse empty array with whitespace");
        assert_eq!(range.start(), TextSize::new(0));
        assert_eq!(range.end(), TextSize::new(5));
        assert_eq!(items_count, 0);
    }

    #[test]
    fn array_with_whitespace() {
        bind_context!(let context = "[ 1 , 2 , 3 ]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let mut items = Vec::new();
        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |state| {
            let token = state.advance().expect("should have at least one token");
            items.push(token.kind.syntax());
            Ok(())
        });

        let range = result.expect("should be able to parse array with whitespace");
        assert_eq!(range.start(), TextSize::new(0));
        assert_eq!(range.end(), TextSize::new(13));
        assert_eq!(
            items,
            [SyntaxKind::Number, SyntaxKind::Number, SyntaxKind::Number]
        );
    }

    #[test]
    fn nested_arrays() {
        bind_context!(let context = "[[1, 2], [3, 4]]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let mut outer_items = Vec::new();
        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |state| {
            let token = state.advance().expect("should have at least one token");
            outer_items.push(token.kind.syntax());

            if token.kind.syntax() == SyntaxKind::LBracket {
                let mut inner_items = Vec::new();
                let inner_result: Result<_, ArrayDiagnostic> = visit_array(state, token, |state| {
                    let token = state.advance().expect("should have at least one token");
                    inner_items.push(token.kind.syntax());
                    Ok(())
                });

                inner_result.expect("should be able to parse nested array");
                assert!(!inner_items.is_empty(), "Nested array should have items");
            }

            Ok(())
        });

        let range = result.expect("should be able to parse nested arrays");
        assert_eq!(range.start(), TextSize::new(0));
        assert_eq!(range.end(), TextSize::new(16));
        assert_eq!(outer_items, [SyntaxKind::LBracket, SyntaxKind::LBracket]);
    }

    #[expect(clippy::panic_in_result_fn)]
    fn process_nested_array(
        state: &mut ParserState<'_, '_>,
        depth: &mut usize,
    ) -> Result<(), ArrayDiagnostic> {
        *depth += 1;

        let token = state.advance().expect("should have at least one token");

        if token.kind.syntax() == SyntaxKind::LBracket && *depth < 3 {
            visit_array(state, token, |state| process_nested_array(state, depth)).map(|_| ())
        } else {
            assert_eq!(token.kind.syntax(), SyntaxKind::Number);
            Ok(())
        }
    }

    #[test]
    fn deep_nesting() {
        bind_context!(let context = "[[[1]]]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let mut depth = 0;

        let result = visit_array(&mut state, token, |state| {
            process_nested_array(state, &mut depth)
        });

        let range = result.expect("should be able to parse deeply nested arrays");
        assert_eq!(range.start(), TextSize::new(0));
        assert_eq!(range.end(), TextSize::new(7));
        assert_eq!(depth, 3);
    }

    #[test]
    fn empty_nested_arrays() {
        bind_context!(let context = "[[], []]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let mut outer_callback_count = 0;
        let mut inner_callback_count = 0;

        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |state| {
            let token = state.advance().expect("should have at least one token");
            assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);
            outer_callback_count += 1;

            visit_array(state, token, |_| {
                inner_callback_count += 1;
                Ok::<(), ArrayDiagnostic>(())
            })
            .expect("should be able to parse empty nested array");

            Ok(())
        });

        let range = result.expect("should be able to parse empty nested arrays");
        assert_eq!(range.start(), TextSize::new(0));
        assert_eq!(range.end(), TextSize::new(8));
        assert_eq!(outer_callback_count, 2);
        assert_eq!(inner_callback_count, 0);
    }

    #[test]
    fn array_with_objects() {
        // This test only checks that object tokens are passed correctly,
        // not that objects are properly parsed (that would be handled by a different function)
        bind_context!(let context = "[{}, {\"key\": 42}]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let mut items = Vec::new();
        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |state| {
            let token = state.advance().expect("should have at least one token");
            items.push(token.kind.syntax());

            loop {
                let token = state.advance().expect("should have at least one token");
                items.push(token.kind.syntax());

                if token.kind.syntax() == SyntaxKind::RBrace {
                    break;
                }
            }

            Ok(())
        });

        let range = result.expect("should be able to parse array with objects");
        assert_eq!(range.start(), TextSize::new(0));
        assert_eq!(range.end(), TextSize::new(17));
        assert_eq!(
            items,
            [
                SyntaxKind::LBrace,
                SyntaxKind::RBrace,
                SyntaxKind::LBrace,
                SyntaxKind::String,
                SyntaxKind::Colon,
                SyntaxKind::Number,
                SyntaxKind::RBrace
            ]
        );
    }

    #[test]
    fn multiline_array() {
        bind_context!(let context = "[\n  1,\n  2,\n  3\n]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let mut items = Vec::new();
        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |state| {
            let token = state.advance().expect("should have at least one token");
            items.push(token.kind.syntax());
            Ok(())
        });

        let range = result.expect("should be able to parse multiline array");
        assert_eq!(range.start(), TextSize::new(0));
        assert_eq!(range.end(), TextSize::new(17));
        assert_eq!(
            items,
            [SyntaxKind::Number, SyntaxKind::Number, SyntaxKind::Number]
        );
    }

    #[test]
    fn consecutive_commas() {
        bind_context!(let context = "[1,,2]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |state| {
            state.advance().expect("should have at least one token");
            Ok(())
        });

        let diagnostic =
            result.expect_err("should not be able to parse array with consecutive commas");
        let report = render_diagnostic(context.input, diagnostic, &context.spans);

        with_settings!({
            description => "Array with consecutive commas should fail"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, report, &context.input);
        });
    }

    #[test]
    fn multiple_consecutive_commas() {
        bind_context!(let context = "[1,,,2]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |state| {
            state.advance().expect("should have at least one token");
            Ok(())
        });

        let diagnostic =
            result.expect_err("should not be able to parse array with multiple consecutive commas");
        let report = render_diagnostic(context.input, diagnostic, &context.spans);

        with_settings!({
            description => "Array with multiple consecutive commas should fail"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, report, &context.input);
        });
    }

    #[test]
    fn consecutive_trailing_commas() {
        bind_context!(let context = "[1,,]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |state| {
            state.advance().expect("should have at least one token");
            Ok(())
        });

        let diagnostic =
            result.expect_err("should not be able to parse array with consecutive commas");
        let report = render_diagnostic(context.input, diagnostic, &context.spans);

        with_settings!({
            description => "Array with consecutive commas should fail"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, report, &context.input);
        });
    }

    #[test]
    fn multiple_leading_commas() {
        bind_context!(let context = "[,,1]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |state| {
            state.advance().expect("should have at least one token");
            Ok(())
        });

        let diagnostic = result.expect_err("should fail with leading comma");
        let report = render_diagnostic(context.input, diagnostic, &context.spans);

        with_settings!({
            description => "Array with multiple leading commas should fail"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, report, &context.input);
        });
    }

    #[test]
    fn leading_comma() {
        bind_context!(let context = "[,1]");
        bind_state!(let mut state from context);

        let token = state.advance().expect("should have at least one token");
        assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

        let result: Result<_, ArrayDiagnostic> = visit_array(&mut state, token, |state| {
            state.advance().expect("should have at least one token");
            Ok(())
        });

        let diagnostic = result.expect_err("should fail with leading comma");
        let report = render_diagnostic(context.input, diagnostic, &context.spans);

        with_settings!({
            description => "Array with leading comma should fail"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, report, &context.input);
        });
    }
}
