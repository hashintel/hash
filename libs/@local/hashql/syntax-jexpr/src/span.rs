use hashql_core::span::{SpanId, TextRange};
use hashql_diagnostics::span::DiagnosticSpan;
use jsonptr::PointerBuf;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Span {
    pub range: TextRange,
    pub pointer: Option<PointerBuf>,

    pub parent_id: Option<SpanId>,
}

impl Span {
    pub(crate) const fn new(range: TextRange) -> Self {
        Self {
            range,
            pointer: None,
            parent_id: None,
        }
    }
}

impl hashql_core::span::Span for Span {
    fn parent_id(&self) -> Option<SpanId> {
        self.parent_id
    }
}

impl From<&Span> for DiagnosticSpan {
    fn from(value: &Span) -> Self {
        Self {
            range: value.range,
            parent_id: value.parent_id,
        }
    }
}
