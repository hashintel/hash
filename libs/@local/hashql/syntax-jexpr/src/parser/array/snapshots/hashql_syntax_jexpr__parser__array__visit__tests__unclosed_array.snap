---
source: libs/@local/hashql/syntax-jexpr/src/parser/array/visit.rs
description: Unclosed array should fail
expression: "[1, 2"
---
[31m[lexer::lexer::unexpected-eof] Error:[0m Lexer
   â•­â”€[ <unknown>:1:6 ]
   â”‚
 1 â”‚ [1, 2
   â”‚      â”‚ 
   â”‚      â•°â”€ Unexpected end of file
   â”‚ 
   â”‚ Help: Missing closing bracket. Make sure all opening brackets have matching closing brackets.
â”€â”€â”€â•¯
